libcst/metadata/full_repo_manager.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+
+import json
+import subprocess
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+)
+
+from mypy_extensions import TypedDict
+
+import libcst as cst
+from libcst.metadata.type_inference_provider import (
+    InferredType,
+    PyreData,
+    TypeInferenceProvider,
+)
+from libcst.metadata.wrapper import MetadataWrapper
+
+
+if TYPE_CHECKING:
+    from libcst.metadata.base_provider import ProviderT
+
+
+class FullRepoManager:
+    def __init__(
+        self,
+        repo_root_dir: str,
+        paths: List[str],
+        providers: List["ProviderT"],
+        timeout: int = 5,
+    ) -> None:
+        self.root_path: Path = Path(repo_root_dir)
+        self._cache: Dict["ProviderT", Mapping[str, object]] = {}
+        self._timeout = timeout
+        for provider in providers:
+            handler = self._get_cache_handler(provider)
+            if handler:
+                self._cache[provider] = handler(paths)
+
+    def _get_cache_handler(self, provider: "ProviderT") -> Optional[Callable]:
+        maps: Dict["ProviderT", Callable[[List[str]], Mapping[str, object]]] = {
+            TypeInferenceProvider: self._handle_pyre_cache
+        }
+        return maps.get(provider)
```

Comment:
Should we make this map an argument to the constructor instead?  I notice that in your tests that this method is patched.  If we make the maps an argument, then we can avoid patching and pass in our own handler method.

Reply:
This method is not patched and it's run in the tests.
The patched method is `_handle_pyre_cache` which it requires actually query pyre daemon (which is an integration test). We don't want to build an integration test as a unit test to avoid flaky or slow test.

Reply:
Thanks for moving the expensive calls out!  This definitely feels better.  I think I wasn't as clear on the other part before.  I'm wondering if we should pass in the map you define on line 65 into the constructor as well.  This way, we can pass in a lambda function as the handler function for `TypeInferenceProvider` instead of mocking out `_handle_pyre_cache`.  In the long term, this feels to me so that we don't bloat this class with handler function implementations, but you've thought about this API design a bit more than I have.

Reply:
The current implementation keeps `TypeInferenceProvider` shell out logic in FullRepoManager. The other way is to keep the logic in `TypeInferenceProvider` to keep related logic together.